Deploy contract while changing beneficiary

sumitnitsurat opened the pull request:
- functionality to deploy contract when endorsing holder and beneficiary
- handle deploying contract when no endorsed user type, 
 - display endorsed beneficiary and holder and change owner to new escrow contract address

Commit:
rebase branch

Commit:
rebase branch

rjchow commented on src/components/AssetInfo/TokenSideBarHolder.test.tsx:
```diff
@@ -16,15 +17,16 @@ describe("tokenSideBarHolder", () => {
     );
-    const inputInstance = wrapper.find("#sec-changebeneficiary input").instance() as any;
-    expect(inputInstance.value).toEqual("");
+    // const inputInstance = wrapper.find("#sec-changebeneficiary input").instance() as any;
```
👎🏻

sumitnitsurat replied:
fixed

rjchow commented on src/components/AssetInfo/TokenSideBarHolder.tsx:
```diff
@@ -26,16 +27,17 @@ const isSurrenderDocumentError = (error: any): error is ErrorType =>
-  const showChangeBeneficiary = !!approvedBeneficiaryAddress || isEqualBeneficiaryAndHolder;
-  const isApprovedBeneficiaryAddress = !!approvedBeneficiaryAddress && !isEqualBeneficiaryAndHolder;
-
+  const isApprovedAddress = !!approvedBeneficiaryAddress && !!approvedHolderAddress;
```
wasn't this logic extracted out?

sumitnitsurat replied:
we extracted out the logic to get the role i.e beneficiary, holder to display the sidebar content and name. this is just to check for the approved addresses and make the input field disable or not based on that.

rjchow commented on src/services/token/token.js:
```diff
@@ -52,3 +53,23 @@ export const endorseBeneficiaryTransfer = async newBeneficiary => {
+
+export const deployEscrowContract = async ({ registryAddress, beneficiaryAddress, holderAddress }) => {
+  const { provider, signer } = await getProvider();
+  const contractOnwerInstance = await WriteableTitleEscrowOwner.deployEscrowContract({
```
variable name spelling

sumitnitsurat replied:
fixed

rjchow commented on src/components/AssetInfo/TokenSideBarContent.tsx:
```diff
@@ -54,9 +68,16 @@ const TokenSideBarContent = ({
+        approvedContractAddress: approvedEscrowContractAddress || ""
+      }
+    });
     // eslint-disable-next-line react-hooks/exhaustive-deps
```
this doesn't seem to be a rule that should be disabled?

sumitnitsurat replied:
I am using useEffect as a componentWillRecieve life cycle method. In the case it is adding fieldValue to the array which is cauing it to go into inifinite loop. find this way of doing it online though will surely search more about it.

Nebulis replied:
You mean you use a hook to behave like a deprecated method ? (https://reactjs.org/docs/react-component.html#unsafe_componentwillreceiveprops)

What's the reason of putting the props into a local state ? looks like you can directly use the props and get rid of all the noise (sync props => state, useEffect, props/state dont have the same name)

sumitnitsurat replied:
this is an input field state and there can be an initial value of it in some cases. so cant mutate the props..so setting the state and mutating.

Nebulis replied:
there is a problem with the flow:
- initial value comes from a very high parent component (from redux)
- this component holds a state based on the parent (never a good idea)
- the state is then updated by child component using an indirect link => `setFieldValue({ ...fieldValue, ...{ [e.target.name]: e.target.value } });`, the name of the child component form must match the name of the parent component field state.

I would suggest to create separate state instead of holding everything into an object, with separate functions to update (why passing each value down individually and not as an object, but only one function ?).
I would also suggest to create a context somewhere to avoid those intermediate state and have one single source of truth for this, because for the moment it's widely split.

Additional question => In which scenario the props is likely to change ?

sumitnitsurat replied:
it's an async call to fetch the holder and beneficiary value. So the initial state value is an empty string. and when the value is fetched from the blockchain then the props get updated.

sumitnitsurat replied:
Also would it be better to use useSelector to access the state in component instead of context ?

Nebulis replied:
Maybe shouldnt display the component while it's loading ?

I dont mind context or selector as long as the state is not split into multiple components with indirect coupling.

sumitnitsurat replied:
yea need to put loader while fetching data. have to extract out loader code in a separate component also. will work on another PR. already made an issue for that. made some changes for state refactoring in latest commit. Please check.

rjchow commented on src/components/AssetInfo/TokenSideBarContent.tsx:
```diff
@@ -78,19 +99,42 @@ const TokenSideBarContent = ({
-  const approveChangeBeneficiary = () => {
-    const { approvedBeneficiary } = fieldValue;
-    handleFormActions(endorseTransfer, TOKEN_ACTION_TYPES.ENDORSE_BENEFICIARY, approvedBeneficiary);
+  const deployEscrowContractAction = async () => {
```
should you be doing this here?

sumitnitsurat replied:
its just used to toggle local state of loaders and error while calling the the deployContract service.

rjchow reviewed (commented):
in general I don't like the sprawl, please focus on refactoring it next sprint to extract all the async/token logic into a useToken hook or something like that

Commit:
pr changes

- refactored getApprovedUsers method. Add promise.all.

Commit:
pr changes

- refactored getApprovedUsers method. Add promise.all.

Commit:
PR changes: breakdown state object into seprate states

- using useSelecgtor to access state directly in child component

Commit:
PR changes: breakdown state object into seprate states

- using useSelecgtor to access state directly in child component

simboonlong committed:
chore: remove bootstrap cdn, use @imports from package.json (#79)

chore: upgrade bootstrap scripts dependancies too

simboonlong committed:
refactor: group scss into abstracts (#81)

chore: remove font import from head, use scss import and only specific font-weights

rjchow committed:
fix: show only revoked message when status is both issued and revoked

rjchow committed:
fix: show only revoked message when status is both issued and revoked

simboonlong committed:
chore: fix all !important instances, avoid use (#82)

* chore: fix all !important instances, avoid use
chore: cleanup unused styles

* fix: remove static style.css, move into respective scss files
chore: allow use of global static helpers

benjaminlim00 committed:
feat: universal actions (#80)

simboonlong committed:
chore: consolidate all color variables, standardize naming convention (#83)

chore: use from defined font family mixin
chore: load the correct needed font families from 1 place
fix: forgot to remove style.css no longer needed
fix: mobile alignments

rjchow committed:
chore: fix test for revoke message change

rjchow committed:
chore: fix test for revoke message change

rjchow committed:
chore: lint

rjchow committed:
chore: lint

Commit:
rebase
